util/caller/resolver.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package caller
+
+import (
+	"os"
+	"regexp"
+	"runtime"
+	"strings"
+	"sync"
+)
+
+type cachedLookup struct {
+	file string
+	line int
+	fun  string
+}
+
+var dummyLookup = cachedLookup{file: "???", line: 1, fun: "???"}
+
+// A CallResolver is a helping hand around runtime.Caller() to look up file,
+// line and name of the calling function. CallResolver caches the results of
+// its lookups and strips the uninteresting prefix from both the caller's
+// location and name; see NewCallResolver().
+type CallResolver struct {
+	mu    sync.Mutex
+	cache map[uintptr]*cachedLookup
+	re    *regexp.Regexp
+}
+
+// defaultPattern strips src/github.com/organization/project/module/submodule/file.go
+// to module/submodule/file.go.
+var defaultPattern = "src/" + strings.Repeat(strings.Join([]string{"[^", "]+", ""}, string(os.PathSeparator)), 3) + "(.*)"
```

Comment:
> @tamird so you're saying you don't want to truncate at all (or maybe just remove src/)? I think that's too verbose for our "human-readable" log format; repo-relative paths seem to me to strike a good balance here since the vast majority will be under the cockroach repo.

If we don't strip at all, we'll get absolute paths, so at minimum we want to strip `$GOPATH/src/`. As for more stripping, I think we can strip to repo-relative just for cockroach, and leave others untouched without complicating this code much.

> Alternately, if we want to avoid any ambiguity with repo-relative paths (because many repos may have a util subpackage), we could map the common packages to short abbreviations

This means none of these will be clickable in terminal, which is a pretty nice feature.